### PR TITLE
Rewrite of Reference Transformations

### PR DESCRIPTION
**Pull request by pgruetter**
### What changes are included in the pull request?
Rewrite of Reference Page "Transformations"

### CODE REVIEW

File: docs/reference/transformations.md [hunk 1]
```diff
@@ -3,158 +3,240 @@ id: transformations
...
+SDLB comes with commonly used transformers out-of-the-box.
 
-:::warning Deprecation Warning
- there has been a refactoring of transformations in Version 2.0.5. The attribute **transformer** is deprecated and will be removed in future versions. Use **transformers** instead.
-:::
+The following predefined 1-to-1 transformations are supported:
 
+| Transformer | Description                                                                                           |
+| ----------- |-------------------------------------------------------------------------------------------------------|
+| AdditionalColumnsTransformer | Add additional columns to the DataFrame by extracting information from the context                    |
```
Comment (zzeekk) [inline docs/reference/transformations.md]:
```suggestion
| AdditionalColumnsTransformer | Add additional columns to the DataFrame by extracting information from the context  or derive new columns from existing columns  |
```

File: docs/reference/transformations.md [hunk 2]
```diff
@@ -3,158 +3,240 @@ id: transformations
...
+| BlacklistTransformer | Apply a column blacklist to discard columns                                                           |
+| DataValidationTransformer | validates DataFrame with a user defined set of rules and creates column with potential error messages |
+| DecryptColumnsTransformer | Decrypts specified columns using AES/GCM algorithm                                                    |
+| EncryptColumnsTransformer | Encrypts specified columns using AES/GCM algorithm                                                    |
+| FilterTransformer | Filter DataFrame with expression                                                                      |
+| SparkRepartitionTransformer | Repartions a DataFrame                                                                                |
```
Comment (zzeekk) [inline docs/reference/transformations.md]:
```suggestion
| SparkRepartitionTransformer | Repartitons a DataFrame                                                                                |
```

File: docs/reference/transformations.md [hunk 3]
```diff
@@ -3,158 +3,240 @@ id: transformations
...
+| StandardizeColNamesTransformer | Used to standardize column names according to configurable rules                                      |
+| StandardizeDatatypesTransformer | Standardize data types of a DataFrame (decimal to corresponding int / float)                          |
```
Comment (zzeekk) [inline docs/reference/transformations.md]:
```suggestion
| StandardizeSparkDatatypesTransformer| Standardize data types of a DataFrame (decimal to corresponding int / float)                          |
```

File: docs/reference/transformations.md [hunk 4]
```diff
@@ -3,158 +3,240 @@ id: transformations
...
+| ScalaClassGenericDfTransformer   | Spark DataFrame transformation in Scala 1-to-1            |
+| ScalaClassGenericDfsTransformer  | Spark DataFrame transformation in Scala many-to-many      |
+| ScalaClassSnowparkDfTransformer  | Snowpark (Snowflake) transformation in Scala 1-to-1       |
+| ScalaClassSnowparkDfsTransformer | Snowpark (Snowflake) transformation in Scala many-to-many |
+| ScalaClassSparkDfTransformer     | Spark DataFrame transformation in Scala 1-to-1            |
+| ScalaClassSparkDfsTransformer    | Spark DataFrame transformation in Scala many-to-many      |
+| ScalaClassSparkDsTransformer     | Spark DataSet transformation in Scala 1-to-1              |
+| ScalaClassSparkDsNTo1Transformer | Spark DataFrame transformation in Scala many-to-one       |
+| ScalaNotebookSparkDfTransformer  | Loads custom code from a Notebook                         | 
+| PythonCodeSparkDfTransformer | Python implementation (using PySpark)                     |
```
Comment (zzeekk) [inline docs/reference/transformations.md]:
```suggestion
| PythonCodeSparkDfTransformer | Spark DataFrame transformation in Python 1-to-1 (using PySpark)  |
| PythonCodeSparkDfsTransformer | Spark DataFrame transformation in Python many-to-many (using PySpark) |
```

File: docs/reference/transformations.md [hunk 5]
```diff
@@ -3,158 +3,240 @@ id: transformations
...
-#### Scala/Java
-In general, Scala/Java transformations can be provided within the configuration file or in seperate source files. 
-You can use Spark Dataset API in Java/Scala to define custom transformations.
-If you have a Java project, create a class that extends CustomDfTransformer or CustomDfsTransformer and implement `transform` method.
-Then use **type = ScalaClassSparkDfTransformer** or **type = ScalaClassSparkDfsTransformer** and configure **className** attribute.
+### SQL
+Spark SQL is probably the easiest way to write a custom transformation, directly in your HOCON configuration.
+All input Data Objects are available as tables to select from.
+Because of naming limitations, all special characters are replaced by an underscore. 
+So an input Data Object called `Table-With-Hyphen` becomes `table_with_hyphen` inside the SQL query.
```
Comment (zzeekk) [inline docs/reference/transformations.md]:
```suggestion
All input Data Objects are available in the select statement with following naming:
- special characters are replaced by an underscore
- a postfix `_sdltemp` is added.
So an input Data Object called `table-with-hyphen` becomes `table_with_hyphen_sdltemp` inside the SQL query.
To simplify this you can also use the special token %{inputViewName} for 1-to-1 transformations, or ${inputViewName_<inputDataObjectName>} for n-to-m transformations, that will be replaced with the correct name at runtime.
```

File: docs/reference/transformations.md [hunk 6]
```diff
@@ -3,158 +3,240 @@ id: transformations
...
-        (session: SparkSession, options: Map[String,String], df: DataFrame, dataObjectId: String) => {
-          import session.implicits._
-          df.select("col1","col2")
-        }
-      """
-    }
-    ...
+transformers = [{
+  type = SQLDfTransformer
+  sqlCode = "select id, count(*) from dataObject1 group by id"
```
Comment (zzeekk) [inline docs/reference/transformations.md]:
```suggestion
  sqlCode = "select id, count(*) from %{inputViewName} group by id"
```

File: docs/reference/transformations.md [hunk 7]
```diff
@@ -3,158 +3,240 @@ id: transformations
...
-          val df_res = df_in1.join(df_in2, $"tbl1_id" === $"tbl2_id", "left").drop("tbl2_id")
-        Map("int-tbl12" -> df_res)             // map output DataFrame
-        }
-      """
-    }]
+transformers = [{
+  type = SQLDfsTransformer
+  code = {
+    dataObjectOut1 = "select id, cnt from dataObjectIn1 where group = 'test1'",
+    dataObjectOut2 = "select id, cnt from dataObjectIn1 where group = 'test2'"
```
Comment (zzeekk) [inline docs/reference/transformations.md]:
```suggestion
    dataObjectOut1 = "select id, cnt from %{inputViewName_dataObjectIn1} where group = 'test1'",
    dataObjectOut2 = "select id, cnt from %{inputViewName_dataObjectIn1} where group = 'test2'"
```

File: docs/reference/transformations.md [hunk 8]
```diff
@@ -3,158 +3,240 @@ id: transformations
...
+        df.select("col1","col2")
+      }
+    """
+  }]
+```
+Check the method signature in CustomDfTransformer and CustomDfsTransformer according to your requirements.
+
+
+### Python
+At the moment, only 1-to-1 transformations are supported in Python.
```
Comment (zzeekk) [inline docs/reference/transformations.md]: now also n-to-m :-)

File: docs/reference/transformations.md [hunk 9]
```diff
@@ -168,14 +250,51 @@ transformers = [{
...
+* `runtimeOptions`: extracted at runtime from the context. 
+
+
+
+##### In SQL
+If you want to use runtimeOptions in SQL, the syntax is %{key}:
+```
+transformers = [{
+  type = SQLDfTransformer
+  sqlCode = "select id, cnt, '%{test}' as test, %{run_id} as last_run_id from dataObject1"
```
Comment (zzeekk) [inline docs/reference/transformations.md]:
```suggestion
  sqlCode = "select id, cnt, '%{test}' as test, %{last_run_id} as last_run_id from dataObject1"
```

File: docs/reference/transformations.md [hunk 10]
```diff
@@ -168,14 +250,51 @@ transformers = [{
...
+Code written can probably only be used for one specific action. 
+
+For custom transformers, you can therefore provide additional options:
+* `options`: static options provided in your HOCON configuration
+* `runtimeOptions`: extracted at runtime from the context. 
+
+
+
+##### In SQL
+If you want to use runtimeOptions in SQL, the syntax is %{key}:
```
Comment (zzeekk) [inline docs/reference/transformations.md]:
```suggestion
If you want to use options in SQL, the syntax is %{key}:
```

File: docs/reference/transformations.md [hunk 11]
```diff
@@ -3,158 +3,240 @@ id: transformations
...
+In custom Scala code, the whole Spark Dataset API is available. 
+It's of course also possible to include additional libraries for your code, 
+so anything you can do in Spark Scala, you can do with SDLB. 
 
-#### SQL
-You can use Spark SQL to define custom transformations.
-Input DataObjects are available as tables to select from. Use tokens %{&ltkey&gt} to replace with runtimeOptions in SQL code.
-For a 1-to-1 transformation use **type = SQLDfTransformer** and configure **code** as your SQL transformation statement.
-For many-to-many transformations use **type = SQLDfsTransformer** and configure **code** as a Map of "&ltoutputDataObjectId&gt, &ltSQL transformation statement&gt".
+##### As separate file
```
Comment (zzeekk) [inline docs/reference/transformations.md]:
```suggestion
##### As Scala class
```

File: docs/reference/transformations.md [hunk 12]
```diff
@@ -3,158 +3,240 @@ id: transformations
...
+If you have a Java/Scala project, it usually makes sense to create separate files for your custom Scala code.
+Any files in your classpath are picked up and can be referenced.
```
Comment (zzeekk) [inline docs/reference/transformations.md]:
```suggestion
If you have a Java/Scala project, it usually makes sense to create separate classes for your custom Scala code.
Any classes in your classpath are picked up and can be referenced. 
```

File: docs/reference/transformations.md [hunk 13]
```diff
@@ -168,14 +250,51 @@ transformers = [{
...
+`sc`, `session` or `sqlContext`. 
+Some additional variables are also available:
+* `inputDf`: Input DataFrame
+* `dataObjectId`: Id of input DataObject as string
+
+The output DataFrame needs to be set with `setOutputDf(df)`.
+
+##### Requirements
+Running Python transformations needs some additional setup. 
+In general, Python >= 3.4 is required and PySpark package needs to be installed with a version matching your SDL spark version.
```
Comment (zzeekk) [inline docs/reference/transformations.md]:
```suggestion
In general, Python >= 3.4 is required and PySpark package needs to be installed with a version matching your SDL spark version. Further environment variable PYTHONPATH needs to be set to your python environment `.../Lib/site-packages` directory, and pyspark command needs to be accessible from the PATH environment variable.
```

**Review by zzeekk — COMMENTED** [review]
Much clearer, thanks. Made some corrections and clarifications to review.

File: docs/reference/transformations.md (hunk 8 above)
Comment (pgruetter) [reply docs/reference/transformations.md]: Right, just in time for this review :-)
Changed the text in https://github.com/smart-data-lake/smart-data-lake/pull/667/commits/a820df6a0ccafeb0e1eb702cca732eb0f264fac6. 

Not too sure about the syntax. Would that be a valid example?
```
transformers = [{
  type = PythonCodeDfsTransformer 
  code = """
    |dfResult = inputDfs[df1].join(inputDfs[df2], inputDfs[df1].id == inputDfs[df2].fk, "left")\
    | .select(inputDfs[df1].id, inputDfs[df1].desc, inputDfs[df2].desc)\
    |setOutputDf("outputId"->dfResult)
  """
}]
```

Comment (zzeekk) [reply docs/reference/transformations.md]:
```
transformers = [{
  type = PythonCodeDfsTransformer 
  code = """
    |dfResult = inputDfs[df1].join(inputDfs[df2], inputDfs[df1].id == inputDfs[df2].fk, "left") \
    |  .select(inputDfs[df1].id, inputDfs[df1].desc, inputDfs[df2].desc)
    |setOutputDf({"outputId": dfResult})
  """
}]
```

Comment (pgruetter) [reply docs/reference/transformations.md]: Thanks a lot. Added the example and description, please re-check this part: https://github.com/smart-data-lake/smart-data-lake/pull/667/commits/62137a5d6c43661d8c2908db6e5e85e86a4016d3

File: docs/reference/transformations.md [hunk 14]
```diff
@@ -168,14 +252,65 @@ transformers = [{
...
+  type = PythonCodeDfsTransformer 
+  code = """
+    |dfResult = inputDfs[df1].join(inputDfs[df2], inputDfs[df1].id == inputDfs[df2].fk, "left") \
+    |  .select(inputDfs[df1].id, inputDfs[df1].desc, inputDfs[df2].desc)
+    |setOutputDf({"outputId": dfResult})
+  """
+}]
+```
+
+In this case, `inputDfs` is an array and `setOutputDf` expects a map.
```
Comment (zzeekk) [inline docs/reference/transformations.md]:
```suggestion
In this case, `inputDfs` is a dictionary (dict) and `setOutputDf` expects a dictionary.
```

File: docs/reference/transformations.md [hunk 15]
```diff
@@ -168,14 +252,65 @@ transformers = [{
...
+##### In Scala
+If you check the signature of the `transform` method again, you can see that you get more than just the DataFrames
+you can manipulate.
+You also get a Map[String,String] called `options`. 
+This Map contains the combined `options` and `runtimeOptions`. 
+So in your custom class, you can read all options and runtimeOptions and use them accordingly to parametrize your code.
 
-How it works: under the hood a PySpark DataFrame is a proxy for a Java Spark DataFrame. PySpark uses Py4j to access Java objects in the JVM.
+##### In Python
+Similarly in Python, in addition to the variables `inputDf` and `dataObjectId`, you get a variable called `options`
```
Comment (zzeekk) [inline docs/reference/transformations.md]:
```suggestion
Similarly in Python, in addition to the variables `inputDf` and `dataObjectId` (resp. `inputsDfs'), you get a variable called `options`
```

**Review by zzeekk — APPROVED** [review]

**Pull request merged**